### PR TITLE
Support origin regex

### PR DESCRIPTION
**Pull request by somlor**
This adds support for regex in origin. (See https://github.com/mschae/cors_plug/pull/21 for comments re: tweaks to tests).

### CODE REVIEW

File: test/cors_plug_test.exs [hunk 1]
```diff
@@ -14,7 +14,8 @@ defmodule CORSPlugTest do
 
   test "lets me overwrite options" do
     opts = CORSPlug.init(origin: "example.com")
-    conn = conn(:get, "/", nil, headers: [{"origin", "example.com"}])
+    conn = conn(:get, "/")
+    |> put_req_header("origin", "example.com")
```
Comment (mschae) [inline test/cors_plug_test.exs]: let's use the more standard way of indenting multi line pipes:

```Elixir
conn =
  conn(:get, "/")
  |> put_req_header("origin", "example.com")
```

File: test/cors_plug_test.exs [hunk 2]
```diff
@@ -42,27 +43,67 @@ defmodule CORSPlugTest do
     end
   end
 
-  test "returns the origin when it is valid" do
-    opts = CORSPlug.init(origin: ["example1.com", "example2.com"])
-    conn = conn(:get, "/", nil, headers: [{"origin", "example1.com"}])
+  test "returns the origin when equal to origin string" do
+    opts = CORSPlug.init(origin: "example1.com")
+    conn = conn(:get, "/")
+    |> put_req_header("origin", "example1.com")
```
Comment (mschae) [inline test/cors_plug_test.exs]: See above

File: test/cors_plug_test.exs [hunk 3]
```diff
@@ -42,27 +43,67 @@ defmodule CORSPlugTest do
...
     conn = CORSPlug.call(conn, opts)
     assert assert ["example1.com"] ==
            get_resp_header(conn, "access-control-allow-origin")
   end
 
-  test "returns null string when the origin is invalid" do
+  test "returns null string when not equal to origin string" do
+    opts = CORSPlug.init(origin: "example1.com")
+    conn = conn(:get, "/")
+    |> put_req_header("origin", "example2.com")
```
Comment (mschae) [inline test/cors_plug_test.exs]: See above

File: test/cors_plug_test.exs [hunk 4]
```diff
@@ -42,27 +43,67 @@ defmodule CORSPlugTest do
...
+
+    conn = CORSPlug.call(conn, opts)
+    assert ["null"] == get_resp_header conn, "access-control-allow-origin"
+  end
+
+  test "returns the origin when in origin list" do
+    opts = CORSPlug.init(origin: ["example1.com", "example2.com"])
+    conn = conn(:get, "/")
+    |> put_req_header("origin", "example2.com")
```
Comment (mschae) [inline test/cors_plug_test.exs]: See above

File: test/cors_plug_test.exs [hunk 5]
```diff
@@ -42,27 +43,67 @@ defmodule CORSPlugTest do
...
+    conn = CORSPlug.call(conn, opts)
+    assert assert ["example2.com"] ==
+           get_resp_header(conn, "access-control-allow-origin")
+  end
+
+  test "returns null string when not in origin list" do
     opts = CORSPlug.init(origin: ["example1.com"])
-    conn = conn(:get, "/", nil, headers: [{"origin", "example2.com"}])
+    conn = conn(:get, "/")
+    |> put_req_header("origin", "example2.com")
```
Comment (mschae) [inline test/cors_plug_test.exs]: See above

File: test/cors_plug_test.exs [hunk 6]
```diff
@@ -42,27 +43,67 @@ defmodule CORSPlugTest do
...
+
+    conn = CORSPlug.call(conn, opts)
+    assert ["null"] == get_resp_header conn, "access-control-allow-origin"
+  end
+
+  test "returns the origin when matches origin regex" do
+    opts = CORSPlug.init(origin: ~r/^example.+\.com$/)
+    conn = conn(:get, "/")
+    |> put_req_header("origin", "example42.com")
```
Comment (mschae) [inline test/cors_plug_test.exs]: See above

File: test/cors_plug_test.exs [hunk 7]
```diff
@@ -42,27 +43,67 @@ defmodule CORSPlugTest do
...
+
+    conn = CORSPlug.call(conn, opts)
+    assert assert ["example42.com"] ==
+           get_resp_header(conn, "access-control-allow-origin")
+  end
+
+  test "returns null string when does not match origin regex" do
+    opts = CORSPlug.init(origin: ~r/^example.+\.com$/)
+    conn = conn(:get, "/")
+    |> put_req_header("origin", "null-example42.com")
```
Comment (mschae) [inline test/cors_plug_test.exs]: See above

File: test/cors_plug_test.exs [hunk 8]
```diff
@@ -42,27 +43,67 @@ defmodule CORSPlugTest do
...
     conn = CORSPlug.call(conn, opts)
     assert ["null"] == get_resp_header conn, "access-control-allow-origin"
   end
 
   test "returns the request host when origin is :self" do
     opts = CORSPlug.init(origin: [:self])
-    conn = conn(:get, "/", nil,
-                headers: [{"origin", "http://cors-plug.example"}])
+    conn = conn(:get, "/")
+    |> put_req_header("origin", "http://cors-plug.example")
```
Comment (mschae) [inline test/cors_plug_test.exs]: See above

File: test/cors_plug_test.exs [hunk 9]
```diff
@@ -82,8 +123,8 @@ defmodule CORSPlugTest do
 
   test "allows all incoming headers" do
     opts = CORSPlug.init(headers: ["*"])
-    conn = conn(:options, "/", nil,
-                headers: [{"access-control-request-headers", "custom-header,upgrade-insecure-requests"}])
+    conn = conn(:options, "/")
+    |> put_req_header("access-control-request-headers", "custom-header,upgrade-insecure-requests")
```
Comment (mschae) [inline test/cors_plug_test.exs]: See above

File: test/cors_plug_test.exs (hunk 1 above)
Comment (somlor) [inline test/cors_plug_test.exs]: @mschae thanks! agreed, mind if i also start the pipeline with bare value (e.g., via [elixir style guide](https://github.com/rrrene/elixir-style-guide) and [dogma](https://github.com/lpil/dogma/blob/master/docs/rules.md#pipelinestart)):
```
conn = 
  :get
  |> conn("/")
  |> put_req_header("origin", "example.com")
```

**Comment by mschae** [comment]
Please fix merge conflicts and I'm happy to merge!

**Comment by somlor** [comment]
@mschae fixed 👍

**Comment by mschae** [comment]
Sorry for the delay on this, end-of-year craze. I'll release this in a few minutes.

**Comment by mschae** [comment]
Published in v1.1.3, thanks for your contribution and happy holidays!

**Pull request merged**